server/src/main/java/io/crate/protocols/postgres/ResultSetReceiver.java

```diff
@@ -52,27 +52,49 @@ class ResultSetReceiver extends BaseResultReceiver {
     ResultSetReceiver(String query,
                       DelayableWriteChannel channel,
                       DelayedWrites delayedWrites,
-                      TransactionState transactionState,
                       AccessControl accessControl,
                       List<PGType<?>> columnTypes,
                       @Nullable FormatCodes.FormatCode[] formatCodes) {
         this.query = query;
         this.channel = channel;
         this.delayedWrites = delayedWrites;
         this.directChannel = channel.bypassDelay();
-        this.transactionState = transactionState;
         this.accessControl = accessControl;
         this.columnTypes = columnTypes;
         this.formatCodes = formatCodes;
     }
 
+    /**
+     * Writes the row to the pg channel and flushes the channel if necessary.
+     * The caller should check {@link #isWritable()} after each call, and if it's false, the consumer should be
+     * suspended until the receiver becomes writable again. Otherwise, this can lead to OOM errors as the channel
+     * outbound buffer keeps growing.
+     *
+     * @return a future that is completed once the row was successfully written (flushed)
+     */
     @Override
-    public void setNextRow(Row row) {
+    @Nullable
+    public CompletableFuture<Void> setNextRow(Row row) {
         rowCount++;
-        Messages.sendDataRow(directChannel, row, columnTypes, formatCodes);
-        if (rowCount % 1000 == 0) {
+        ChannelFuture sendDataRow = Messages.sendDataRow(directChannel, row, columnTypes, formatCodes);
+        var future = new CompletableFuture<Void>();
+        sendDataRow.addListener(f -> {
+            if (f.isDone() == false) {
+                return;
+            }
+            if (f.isSuccess()) {
+                future.complete(null);
+            } else {
+                future.completeExceptionally(f.cause());
+            }
+        });
+
+        // Flush the channel only every 1000 rows for better performance.
+        // But flushing must be forced once the channel outbound buffer is full (= channel not in writable state)
+        if (isWritable() == false || rowCount % 1000 == 0) {
             directChannel.flush();
         }
+        return future;
```

Comment:
```suggestion
        ChannelFuture sendDataRow = Messages.sendDataRow(directChannel, row, columnTypes, formatCodes);
        // Flush the channel only every 1000 rows for better performance.
        // But flushing must be forced once the channel outbound buffer is full (= channel not in writable state)
        if (isWritable() == false || rowCount % 1000 == 0) {
            directChannel.flush();
        }
        if (isWriteable() == false) {
            var future = new CompletableFuture<Void>();
            sendDataRow.addListener(f -> {
                if (f.isDone() == false) {
                    return;
                }
                if (f.isSuccess()) {
                    future.complete(null);
                } else {
                    future.completeExceptionally(f.cause());
                }
            });
            return future;
        }
        return null;
```

Could also avoid the future's here for the common case

Reply:
Yeah, missed this initially, but added it by https://github.com/crate/crate/pull/17230/commits/aee44e72a92b0e9062877526130ac8a8fe6cf824

Reply:
Btw. as the `flush` can change the `isWritable`, we must do in only after we registered the listener otherwise the listener calls may have been done already.